### DISCUSSION TITLE
computeChangeSets() removes original persisted changes

### DISCUSSION THREAD

**Discussion by Jcalcaldev** [mikro-orm/mikro-orm#4280] [General]
**Bug description:**

Using computeChangeSets() on an entity subscriber will remove the original changeSet.

I have a User entity with a OneToMany relation to a Session entity. When creating a new session for a user, I want to delete all other user sessions. After scouring the docs and battling for a while, the only way I found for the session delete change sets to go through was using `computeChangeSets()` (PLURAL), but this caused all sessions to be deleted and not the original one to be created. Then I found in the code that computeChangeSets() deletes all previous change sets for a reason I'm not familiar with.

Here is my subscriber:

```
public async handleFlush(cs: ChangeSet<Session>, uow: UnitOfWork) {
    if (cs.type !== ChangeSetType.CREATE) {
      return;
    }

    const { entity: session } = cs;

    const previousSessions = await session.user.sessions.init();

    for (const previousSession of previousSessions) {
      previousSessions.remove(previousSession);
    }

    uow.computeChangeSets();
  }
```

Here is the diff that solved my problem:

```diff
diff --git a/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js b/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js
index 2eae68a..812ace1 100644
--- a/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js
+++ b/node_modules/@mikro-orm/core/unit-of-work/UnitOfWork.js
@@ -331,7 +331,6 @@ class UnitOfWork {
         wrapped.__managed = false;
     }
     computeChangeSets() {
-        this.changeSets.clear();
         const visited = new Set();
         for (const entity of this.removeStack) {
             this.cascade(entity, enums_1.Cascade.REMOVE, visited);
```

**Aditional Context:**
I'm not experienced in MikroOrm as it's my first time using it. Please enlighten me if I'm wrong or if this is actually a bug or an unintended usage of the method.

**Versions:**

Dependency | Version
-- | --
node | 18.14.1
typescript | 4.9.5
mikro-orm | 5.7.2

**Comment by B4nan** [comment]
The method is rather internal, and the cleaning is there on purpose, it is supposed to be called once per flush, by the UoW. 

Looks like we are missing some public API to create delete changesets, that sounds like a better tool for the job, definitely more performant, as `computeChangeSets` is quite costly operation (it goes through all the loaded entities and check them for changes, which already happened by the time you get inside the onFlush event). 

The purpose of `onFlush` hook is to modify the UoW state by hand, not using the EM methods - in other words, you shouldn't be really using `em.remove()` there (or anything that delays the operation to the flush call - as you are already inside the flush call), but rather modify the existing changesets (or add new ones). 

If you care only about the records being deleted, you could do `em.nativeDelete()` instead, that just fires the delete query based on what you pass in the parameters.

**Reply by B4nan** [reply]
Yeah indeed, this is missing, I will add a `type` parameter to the `uow.computeChangeSet` method so it can be used this way too.

For now you could hack it this way:

```ts
async handleFlush(cs: ChangeSet<Session>, uow: UnitOfWork) {
  if (cs.type !== ChangeSetType.CREATE) {
    return;
  }

  const { entity: session } = cs;

  const previousSessions = await session.user.sessions.init();

  for (const previousSession of previousSessions) {
    // @ts-ignore accessing private property
    uow.changeSets.set(previousSession, new ChangeSet(previousSession, ChangeSetType.DELETE, {}, cs.meta));
  }
}
```

**Reply by Jcalcaldev** [reply]
Thanks for the quick reply, both methods worked perfectly.

It gave me a headache but feels awesome to know it will be implemented!

**Reply by B4nan** [reply]
Here you go, dev version will be available on NPM in ~15 minutes

61479b642751f9bc347ae2ab6f0bbf630d6a6295
